src/core/eme/clean_old_loaded_sessions.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright 2015 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  EMPTY,
+  merge as observableMerge,
+  Observable,
+} from "rxjs";
+import {
+  mapTo,
+  startWith,
+} from "rxjs/operators";
+import config from "../../config";
+import { IMediaKeySessionInfo } from "./types";
+import LoadedSessionsStore from "./utils/loaded_sessions_store";
+
+const { EME_MAX_SIMULTANEOUS_MEDIA_KEY_SESSIONS } = config;
+
+/**
+ * Event emitted when an old MediaKeySession has been closed to respect the
+ * maximum limit of concurrent MediaKeySession active.
+ */
+export interface ICleanedOldSessionEvent {
+  type : "cleaned-old-session";
+  value : IMediaKeySessionInfo;
+}
+
+/**
+ * Event emitted when we are beginning to close an old MediaKeySession to
+ * respect the maximum limit of concurrent MediaKeySession active.
+ */
+export interface ICleaningOldSessionEvent {
+  type : "cleaning-old-session";
+  value : IMediaKeySessionInfo;
+}
+
+/**
+ * Close sessions to respect the `EME_MAX_SIMULTANEOUS_MEDIA_KEY_SESSIONS` limit.
+ * Emit event when a MediaKeySession begin to be closed and another when the
+ * MediaKeySession is closed.
+ * @param {Object} loadedSessionsStore
+ * @returns {Observable}
+ */
+export default function cleanOldLoadedSessions(
+  loadedSessionsStore : LoadedSessionsStore
+) : Observable<ICleaningOldSessionEvent | ICleanedOldSessionEvent> {
+  const maxSessions = EME_MAX_SIMULTANEOUS_MEDIA_KEY_SESSIONS;
+  const cleaningOldSessions$ : Array<Observable<ICleanedOldSessionEvent |
+                                                 ICleaningOldSessionEvent>> = [];
+  if (maxSessions > 0 && maxSessions >= loadedSessionsStore.getLength()) {
```

Comment:
```suggestion
  if (maxSessions > 0 && maxSessions < loadedSessionsStore.getLength()) {
```
We start discarding loaded sessions if there are more than `maxSessions`, right ?

Reply:
Oh thanks. Yes, that was a problematic mistake.

Even more considering that this is just a code I moved and the mistake was not here originally (it was `<=` originally, as we will create one just after).

_There's a reason for why that modification appeared here despite being quasi a copy-paste: this feature was known to work, but by re-reading I (wrongly!) thought the code was wrong. As it still worked, I thought that it may be because that was `<=` and not `<` so it would have been right when it became equal. By re-re-reading it I realize that I may need to go more slowly and carefully with those code changes!_

Reply:
Done.

I also realized both that the fact that function tried to make place for a future MediaKeySession was both not really clear just by glancing over it and even not really even functional as the loop which followed did not actually loop one more time.

The responsibility is now moved to the caller, which has to give a `limit` number and I tried to improve this code which was in retrospect not very well-written.
Commit: fcdc9057b979d147eb647bc01dc3453e9ea6acd4

Consequently, I'll add some (unit) tests now.
Thanks again for finding that!

Reply:
Happens all the time :)

Reply:
Totally agree with the new client-specified `limit` approach.